### PR TITLE
Updated surface warp smoothing options

### PR DESCRIPTION
**Pull request by akshaysridhar**
Optional surface diffusion over warped terrain. 
- [x] Update `src/Hypsography/Hypsography.jl`
- [x] Update unit-tests in `test/Spaces/terrain_warp.jl` (Integral, steady-state checks.)

### CODE REVIEW

**Comment by akshaysridhar** [comment]
bors try

**Comment by akshaysridhar** [comment]
bors try

**Comment by akshaysridhar** [comment]
@sriharshakandala I've added the integral test you suggested.

**Comment by akshaysridhar** [comment]
bors try

**Comment by akshaysridhar** [comment]
bors try

**Comment by akshaysridhar** [comment]
bors try

File: src/Hypsography/Hypsography.jl [hunk 1]
```diff
@@ -156,7 +160,46 @@ function ExtrudedFiniteDifferenceSpace(
...
+    # Create dss buffer
+    ghost_buffer = (bf = Spaces.create_dss_buffer(f),)
+    # Apply smoothing
+    for iter in 1:maxiter
+        # Euler steps
+        χf = @. wdiv(grad(f))
+        Spaces.weighted_dss_start!(χf, ghost_buffer.bf)
+        Spaces.weighted_dss_internal!(χf, ghost_buffer.bf)
+        Spaces.weighted_dss_ghost!(χf, ghost_buffer.bf)
+        @. f += κ * dt * χf
```
Comment (sriharshakandala) [inline src/Hypsography/Hypsography.jl:197]: Should we use `ClimaTimeSteppers` here?

File: src/Hypsography/Hypsography.jl [hunk 2]
```diff
@@ -156,7 +160,46 @@ function ExtrudedFiniteDifferenceSpace(
     )
 end
 
-
+"""
+  diffuse_surface_elevation!(f; κ,iter,dt)
```
Comment (simonbyrne) [inline src/Hypsography/Hypsography.jl:164]: Needs 4 spaces for indentation
```suggestion
    diffuse_surface_elevation!(f; κ,iter,dt)
    
```

File: src/Hypsography/Hypsography.jl (hunk 1 above)
Comment (simonbyrne) [reply src/Hypsography/Hypsography.jl:197]: @akshaysridhar would there be any advantage to using a different timestepper here? If so, we can add it in.

File: src/Hypsography/Hypsography.jl (hunk 2 above)
Comment (simonbyrne) [reply src/Hypsography/Hypsography.jl:164]: Also docstring needs to be added to docs, and make sure that the markdown renders as you intend it.

File: src/Hypsography/Hypsography.jl (hunk 1 above)
Comment (akshaysridhar) [reply src/Hypsography/Hypsography.jl:197]: I see DiffEqBase is in the deps here; this is supposed to be a relatively light component in the overall scheme - a Forward-Euler update will suffice (hence the current form). Updating it to use DiffEqBase with new commits to follow soon.

File: src/Hypsography/Hypsography.jl (hunk 2 above)
Comment (akshaysridhar) [reply src/Hypsography/Hypsography.jl:164]: Noted, thanks

**Comment by akshaysridhar** [comment]
bors try

File: src/Hypsography/Hypsography.jl [hunk 3]
```diff
@@ -156,7 +160,46 @@ function ExtrudedFiniteDifferenceSpace(
...
+    FT = eltype(f)
+    # Create dss buffer
+    ghost_buffer = (bf = Spaces.create_dss_buffer(f),)
+    # Apply smoothing
+    for iter in 1:maxiter
+        # Euler steps
+        χf = @. wdiv(grad(f))
+        Spaces.weighted_dss_start!(χf, ghost_buffer.bf)
+        Spaces.weighted_dss_internal!(χf, ghost_buffer.bf)
+        Spaces.weighted_dss_ghost!(χf, ghost_buffer.bf)
```
Comment (sriharshakandala) [inline src/Hypsography/Hypsography.jl:196]:
```
Spaces.weighted_dss_start!(χf, ghost_buffer.bf)
```

**Comment by akshaysridhar** [comment]
cc @simonbyrne @sriharshakandala Code and docs have been updated following your comments.

File: src/Hypsography/Hypsography.jl [hunk 4]
```diff
@@ -156,7 +160,44 @@ function ExtrudedFiniteDifferenceSpace(
...
+elevation to the `HypsographyAdaption` type. Laplacian 
+operator is applied with simple Euler updates to generate
+profiles for each new iteration.
+(1) Compute discrete elevation profile f
+(2) Compute diffuse_surface_elevation!(f, κ, iter). f is mutated.
+(3) Define `Hypsography.LinearAdaption(f)`
+(4) Define `ExtrudedFiniteDifferenceSpace` with new surface elevation.
+Default diffusion parameters are appropriate for spherical arrangements. 
+For `zmax-zsfc` == 𝒪(10^4), κ == 𝒪(10^8), dt == 𝒪(10⁻¹).
+Applicable on periodic domains only.
```
Comment (simonbyrne) [inline src/Hypsography/Hypsography.jl:176]: Doc formatting doesn't look correct, see https://clima.github.io/ClimaCore.jl/previews/PR1262/api/#Hypsography

You can get a live preview using https://github.com/tlienart/LiveServer.jl

**Comment by akshaysridhar** [comment]
bors try

**Comment by akshaysridhar** [comment]
@simonbyrne https://clima.github.io/ClimaCore.jl/previews/PR1262/api/ Seems like the formatting is OK now.

File: test/Spaces/terrain_warp.jl [hunk 5]
```diff
@@ -185,6 +204,50 @@ end
...
+                (zmin, zmax),
+                nh,
+                nl,
+                np;
+                warp_fn = warp_option_2d,
+                test_smoothing = false,
+            )
+            ʷᶜcoords = Fields.coordinate_field(ʷhv_center_space)
+            ʷᶠcoords = Fields.coordinate_field(ʷhv_face_space)
+            ʷᶠʳcoords = Fields.coordinate_field(ʷhv_face_space)
```
Comment (sriharshakandala) [inline test/Spaces/terrain_warp.jl:240]: Should this be from the reference/undiffused surface?

Comment (akshaysridhar) [reply test/Spaces/terrain_warp.jl]: @sriharshakandala Yes, fixed.

File: src/Hypsography/Hypsography.jl [hunk 6]
```diff
@@ -156,7 +160,45 @@ function ExtrudedFiniteDifferenceSpace(
...
+    FT = eltype(f)
+    # Create dss buffer
+    ghost_buffer = (bf = Spaces.create_dss_buffer(f),)
+    # Apply smoothing
+    for iter in 1:maxiter
+        # Euler steps
+        χf = @. wdiv(grad(f))
+        Spaces.weighted_dss!(χf, ghost_buffer.bf)
+        @. f += κ * dt * χf
+        @. f = ifelse(f < FT(0), FT(0), f)
```
Comment (sriharshakandala) [inline src/Hypsography/Hypsography.jl:197]: Do we need this? Depressions are also valid topographical features right?

File: test/Spaces/terrain_warp.jl [hunk 7]
```diff
@@ -185,6 +204,47 @@ end
...
+            ʷᶠcoords = Fields.coordinate_field(ʷhv_face_space)
+            ʷᶠʳcoords = Fields.coordinate_field(ʳhv_face_space)
+            ᶠz₀ = ClimaCore.Fields.level(ʷᶠcoords.z, half)
+            @test minimum(ᶠz₀) >= FT(0)
+            @test maximum(ᶠz₀) <= FT(0.5)
+            @test maximum(@. abs.(ᶠz₀ .- one(ᶠz₀) .* FT.(1 / 4))) <= FT(1e-2)
+        end
+    end
+end
+
```
Comment (sriharshakandala) [inline test/Spaces/terrain_warp.jl:247]: Can we add a 3D example?

File: src/Hypsography/Hypsography.jl (hunk 6 above)
Comment (akshaysridhar) [reply src/Hypsography/Hypsography.jl]: Fro source files where we ignore bathymetric representations; this should be handled at the ClimaAtmos level; so this is unnecessary here. Removed and committed.

**Comment by akshaysridhar** [comment]
@sriharshakandala Updated to use generated extrema, added 3D test, cleaned up names in the warp functions.

**Review by sriharshakandala — APPROVED** [review]

**Pull request merged**